contrib/mesos/pkg/executor/executor.go

```diff
@@ -1044,3 +1002,220 @@ func executorInfoToAnnotations(ei *mesos.ExecutorInfo) (annotations map[string]s
+	// watch pods from the given pod ListWatch
+	if config.PodLW == nil {
+		// fail early to make debugging easier
+		panic("cannot create executor with nil PodLW")
+	}
+
+	maxUpdateBackoff := defaultMaxUpdateBackoff
+	if config.LaunchGracePeriod > 0 && config.LaunchGracePeriod < maxUpdateBackoff {
+		maxUpdateBackoff = config.LaunchGracePeriod / 4
+	}
+	minUpdateBackoff := defaultMinUpdateBackoff
+	if minUpdateBackoff > maxUpdateBackoff {
+		minUpdateBackoff = maxUpdateBackoff / 4
+	}
+	p := &podUpdateProcessor{
+		updateQueue:     queue.NewDelayFIFO(),
+		updateBackoff:   backoff.New(minUpdateBackoff, maxUpdateBackoff),
+		terminate:       terminate,
+		retryExpiration: config.LaunchGracePeriod * 3 / 2,
+	}
+	if p.retryExpiration > 0 {
+		p.updateExpiration = queue.NewDelayFIFO()
+	}
+	_, p.podController = framework.NewInformer(config.PodLW, &api.Pod{}, podRelistPeriod, &framework.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			pod := obj.(*api.Pod)
+			log.V(4).Infof("pod %s/%s created on apiserver", pod.Namespace, pod.Name)
+			qpod := queuer.NewPod(pod, queuer.Deadline(time.Now()))
+			if p.retryExpiration > 0 {
+				p.updateExpiration.Delete(qpod.GetUID())
+			}
+			p.updateQueue.Offer(qpod, queue.ReplaceExisting)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			pod := newObj.(*api.Pod)
+			log.V(4).Infof("pod %s/%s updated on apiserver", pod.Namespace, pod.Name)
+			qpod := queuer.NewPod(pod, queuer.Deadline(time.Now()))
+			if p.retryExpiration > 0 {
+				p.updateExpiration.Delete(qpod.GetUID())
+			}
+			p.updateQueue.Offer(qpod, queue.ReplaceExisting)
+		},
+		DeleteFunc: func(obj interface{}) {
+			pod := obj.(*api.Pod)
+			log.V(4).Infof("pod %s/%s deleted on apiserver", pod.Namespace, pod.Name)
+			uid := queuer.NewPod(pod).GetUID()
+			p.updateQueue.Delete(uid)
+			if p.retryExpiration > 0 {
+				p.updateExpiration.Delete(uid)
+			}
+		},
+	})
+	return p
+}
+
+func (p *podUpdateProcessor) run() {
+	go p.podController.Run(p.terminate)
+	go func() {
+		for {
+			obj := p.updateQueue.Pop(p.terminate)
+			if obj == nil {
+				return
+			}
+			pod := obj.(*queuer.Pod)
+			p.handleChangedApiserverPod(pod)
+		}
+	}()
+	if p.retryExpiration > 0 {
+		go func() {
+			for {
+				obj := p.updateExpiration.Pop(p.terminate)
+				if obj == nil {
+					return
+				}
+				pod := obj.(*queuer.Pod)
+				p.updateQueue.Delete(pod.GetUID())
+			}
+		}()
+	}
+}
+
+func (p *podUpdateProcessor) handleChangedApiserverPod(pod *queuer.Pod) {
+	// exclude "pre-scheduled" pods which have a NodeName set to this node without being scheduled already.
+	// Note: The task-id annotation won't be present on the pod unless the pod has already completed the
+	// launch-task-binding phase. There also could be a race between the overall launch-task process and
+	// this update, but here we have absolutely no way to tell the diff. The best we can do is reschedule
```

Comment:
this update could be executing at the same time as the launch-task flow. consider the following order of events:

1. scheduler sends LaunchTask message w/ pod prototype
1. user updates pod spec
1. executor observes updated pod spec queues an update, eventually invokes this handler
1. mesos slave sends LaunchTask message to executor, bindings invoke LaunchTask on exec

the invocation of this handler won't see the task because the executor hasn't processed the LaunchTask message yet. if we drop the update on the floor we could be missing changes that the user wants to see in the pod.

now that i'm writing this.. it's arguable that the launch-task-binding phase would trigger an update anyway. and that would cause this handler flow to be invoked, at which point we'd know the task-id. and so we'd not need to worry about deferring any earlier updates because we'd get the most recent copy of the pod at that point.

hm.. i thought that I had reasoned out a case where this race could cause problems. am I chasing my tail here?